Automate gcc 5 requirements

seelabs opened the pull request.

nbougalis commented:
Trivial requests, like this, should be marked as "PASSED" by default.

vinniefalco commented:
@nbougalis Not so fast...

@seelabs Should this also be mentioned in the build instructions?

seelabs commented:
@vinniefalco Do you mean here or someplace else? https://wiki.ripple.com/Build_instructions

Are we still supporting 12.04? If so I'll set up a VM tomorrow and run through the instructions and make sure they work.

vinniefalco commented:
@seelabs At least on the wiki.

scottschurr commented:
:+1:

seelabs commented:
@nbougalis @scottschurr I just push a non-trivial change to this PR. I changed `SConstruct` so it detects when it should use the old ABI (works with ubuntu only - an environment variable is still available for other distros). I also added a script to automatically setup a fresh ubuntu system so it's ready to build rippled. I tested the script on ubuntu 12.04, 14.04, 15.04, and 15.10. The 12.04 build will not work out of the box (static linking fails). I believe requiring a custom built boost for this platform will solve this. If these changes are accepted I'll update the wiki with new instructions.

seelabs commented:
I confirmed that static linking now works on ubuntu 12.04 if the user builds boost (a package installed boost works for ubuntu >= 14.04). I added a script to download and build boost with the correct ABI for the various flavours of ubuntu.

scottschurr commented:
LGTM :+1:

MarkusTeufelberger commented:
Please consider parsing /etc/lsb-release instead or using https://docs.python.org/2/library/platform.html#unix-platforms instead of parsing only a single value that exists on non-Debian/Ubuntu distros too and might have very different meaning there.

My gcc already states that by default it will use the GCC4 ABI (gcc -v --> [...] --with-default-libstdcxx-abi=gcc4-compatible) and since there are apparently some issues with GCC and Clang on that subject (http://allanmcrae.com/2015/06/the-case-of-gcc-5-1-and-the-two-c-abis/), my distro might switch over the magic "15.1" value WITHOUT being GCC5 ABI compatible (the "RIPPLED_OLD_GCC_ABI" variable is not commonly set).

All in all, it might work fine for a while and maybe fine for Ubuntu, but this could lead to some trouble for other distros - please make sure your platform specific code (e.g. Ubuntu related hacks only) only targets that platform.

MarkusTeufelberger commented on Builds/Ubuntu/install_boost.sh:
```diff
@@ -0,0 +1,31 @@
+    extra_defines="define=_GLIBCXX_USE_CXX11_ABI=0"
+fi
+
+curl -L "http://downloads.sourceforge.net/project/boost/boost/1.59.0/boost_1_59_0.tar.bz2?r=http%3A%2F%2Fsourceforge.net%2Fprojects%2Fboost%2Ffiles%2Fboost%2F1.59.0%2F&ts=1445463403&use_mirror=iweb" -o boost_1_59_0.tar.bz2
```
Maybe consider downloading from github (https://github.com/boostorg/boost/releases) instead? Your mirror is hardcoded and the transfer is http only...

seelabs replied:
@MarkusTeufelberger Those files don't work for me. `bootstrap.sh` won't run, and the tarballs don't seem to include the sources (they are only ~ 1M and the usual tarball is ~80M). Do you how I install from those tarballs? (I'm guessing they are not meant to be installed from, but I'd love to be wrong).

MarkusTeufelberger replied:
https://svn.boost.org/trac/boost/wiki/TryModBoost#InstallingModularBoost - seems like it is a repository of repositories. Probably there's a way to only load the contents of a certain commit/tag using git.

seelabs commented:
@MarkusTeufelberger 

Thanks. I added checks to make sure the scripts and ABI check logic only run on Ubuntu. If SConstruct is run on non-Ubuntu linux, no ABI flag will be set unless an environment variable is set.

For distros what ship with gcc 5, as long as the library ABI match the compiler ABI, we don't need to do anything special (including if the distro ships a gcc 5 that uses the gcc 4 ABI). The problematic case is when a user installs gcc 5 on a distro that ships with gcc 4. I tried to make that as easy as possible for Ubuntu. Without a lot more work, the best I can offer for other distros is setting the `RIPPLED_OLD_GCC_ABI` flag.

If this turns out to be a large pain point I can look into detecting what the default ABI the gcc compiler is using and check the ABI of boost/protobuf and set flags or issues warning accordingly. I don't know if that's even possible.

seelabs commented:
@MarkusTeufelberger Pushed a commit to build boost from git. Thanks for the suggestion.

MarkusTeufelberger commented:
A pleasure and thanks for reacting so fast, getting as close a possible to the original source is always good! :)

seelabs commented:
@nbougalis Can I get you to take a quick look at the scripts I've added. I'd like to make a new develop candidate soonish so I can reference these scripts on the wiki

nbougalis commented:
On it.

seelabs commented:
@MarkusTeufelberger I'll do what I can to get sources from github. Unfortunately, boost built from github is failing on Ubuntu 12.04 (and boost built from the official tar is succeeding). So there seems to be a difference between what is tagged in git and what is offered in the tarball. I'm looking into this, but wanted to give you heads up in case I have to go back to the old way.

seelabs commented:
Fixed the issue with building boost from git. The problem was I needed up update the submodules. Works on 12.04 now

nbougalis commented:
This seems to work fine for me, on a Ubuntu-derived distribution.

seelabs commented:
Squashed

scottschurr commented:
Looks good as far as I can tell.
:+1:

seelabs commented:
Thanks @scottschurr, took the changes and squashed.
